File: iosp1/P1/P1/UI/Support/FileStats.swift
import Foundation

struct FileReadiness {
    let readinessPercent: Int
    let measurementCount: Int
    let photoCount: Int
    let noteCount: Int
    let expectedMeasurements: Int
    let dayStreak: Int
    let missingHint: String?
    let tagline: String
}

enum FileStats {
    static func compute(followUp: FollowUpUi, events: [TimelineEventResponse]) -> FileReadiness {
        let userEvents = events.filter { $0.type == "user" }
        let measurements = userEvents.filter { !isNoteEvent($0) }
        let measurementCount = measurements.count
        let photoCount = measurements.filter { $0.content.localizedCaseInsensitiveContains("Photo:") }.count
        let noteCount = userEvents.filter { isNoteEvent($0) }.count

        let slotsPerDay = max(followUp.schedule?.count ?? 1, 1)
        let expectedMeasurements = max(followUp.totalDays * slotsPerDay, 1)
        let readinessPercent = min(Int((Float(measurementCount) / Float(expectedMeasurements)) * 100), 100)

        let dayStreak = computeStreak(measurements: measurements)
        let missingHint = buildMissingHint(
            readiness: readinessPercent,
            daysRemaining: followUp.daysRemaining,
            actual: measurementCount,
            expected: expectedMeasurements
        )
        let tagline = dashboardTagline(daysRemaining: followUp.daysRemaining, readiness: readinessPercent)

        return FileReadiness(
            readinessPercent: readinessPercent,
            measurementCount: measurementCount,
            photoCount: photoCount,
            noteCount: noteCount,
            expectedMeasurements: expectedMeasurements,
            dayStreak: dayStreak,
            missingHint: missingHint,
            tagline: tagline
        )
    }

    static func greetingPrefix() -> String {
        let hour = Calendar.current.component(.hour, from: Date())
        if hour < 12 { return "Good morning" }
        if hour < 17 { return "Good afternoon" }
        return "Good evening"
    }

    static func dashboardTagline(daysRemaining: Int, readiness: Int) -> String {
        if daysRemaining == 0 {
            return "Appointment day — your file is ready to share with your doctor."
        }
        if daysRemaining <= 3 && readiness >= 70 {
            return "Almost there. Your doctor will have a clear picture."
        }
        if daysRemaining <= 3 {
            return "Every entry counts — \(daysRemaining) days until your appointment."
        }
        if readiness == 0 {
            return "Start today — your future self will thank you at the appointment."
        }
        if readiness < 40 {
            return "You're building something useful. Keep adding a little each day."
        }
        if readiness < 80 {
            return "Your file is taking shape. Steady progress."
        }
        return "Strong file so far. Stay consistent until appointment day."
    }

    private static func isNoteEvent(_ event: TimelineEventResponse) -> Bool {
        event.date_label.caseInsensitiveCompare("Question") == .orderedSame
    }

    private static func buildMissingHint(
        readiness: Int,
        daysRemaining: Int,
        actual: Int,
        expected: Int
    ) -> String? {
        let remaining = expected - actual
        if daysRemaining == 0 && readiness >= 80 { return nil }
        if daysRemaining == 0 { return "A few more entries will strengthen your file for today." }
        if remaining <= 0 { return nil }
        if readiness >= 90 { return "Almost complete — keep going." }
        if remaining == 1 { return "One more check-in fills a gap in your file." }
        return "\(remaining) check-ins still to capture before your appointment."
    }

    private static func computeStreak(measurements: [TimelineEventResponse]) -> Int {
        guard !measurements.isEmpty else { return 0 }
        let calendar = Calendar.current
        let dates = Set(measurements.compactMap { parseEventDate($0).map { calendar.startOfDay(for: $0) } })
        var streak = 0
        var day = calendar.startOfDay(for: Date())
        while dates.contains(day) {
            streak += 1
            guard let previous = calendar.date(byAdding: .day, value: -1, to: day) else { break }
            day = previous
        }
        return streak
    }

    private static func parseEventDate(_ event: TimelineEventResponse) -> Date? {
        let raw = event.effective_at ?? event.created_at
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: raw) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: raw)
    }
}

File: androidp1/app/src/main/java/com/preappointment1/app/ui/support/FileStats.kt
package com.preappointment1.app.ui.support

import com.preappointment1.app.data.model.TimelineEventResponse
import com.preappointment1.app.ui.screens.FollowUpUi
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId

data class FileReadiness(
    val readinessPercent: Int,
    val measurementCount: Int,
    val photoCount: Int,
    val noteCount: Int,
    val expectedMeasurements: Int,
    val dayStreak: Int,
    val missingHint: String?,
    val tagline: String
)

object FileStats {

    fun compute(followUp: FollowUpUi, events: List<TimelineEventResponse>): FileReadiness {
        val userEvents = events.filter { it.type == "user" }
        val measurements = userEvents.filter { !isNoteEvent(it) }
        val measurementCount = measurements.size
        val photoCount = measurements.count { it.content.contains("Photo:", ignoreCase = true) }
        val noteCount = userEvents.count { isNoteEvent(it) }

        val slotsPerDay = followUp.schedule?.size?.coerceAtLeast(1) ?: 1
        val expectedMeasurements = (followUp.totalDays * slotsPerDay).coerceAtLeast(1)
        val readinessPercent = ((measurementCount.toFloat() / expectedMeasurements) * 100)
            .toInt()
            .coerceIn(0, 100)

        val dayStreak = computeStreak(measurements)
        val missingHint = buildMissingHint(readinessPercent, followUp.daysRemaining, measurementCount, expectedMeasurements)
        val tagline = dashboardTagline(followUp.daysRemaining, readinessPercent)

        return FileReadiness(
            readinessPercent = readinessPercent,
            measurementCount = measurementCount,
            photoCount = photoCount,
            noteCount = noteCount,
            expectedMeasurements = expectedMeasurements,
            dayStreak = dayStreak,
            missingHint = missingHint,
            tagline = tagline
        )
    }

    fun checkInSuccessMessage(daysRemaining: Int, readinessPercent: Int): String = when {
        daysRemaining == 0 -> "Added to your file. Good luck at your appointment today."
        daysRemaining <= 3 && readinessPercent >= 80 ->
            "Noted — thank you. Your briefing is nearly complete ($readinessPercent% ready)."
        daysRemaining <= 3 ->
            "Noted — thank you. $daysRemaining days until your appointment."
        readinessPercent >= 50 ->
            "Saved to your file — $readinessPercent% ready for your doctor."
        else -> "Noted — thank you. We'll take care of the rest."
    }

    fun noteSuccessMessage(): String =
        "Your note was added to the file for your doctor."

    fun greetingPrefix(): String {
        val hour = java.time.LocalTime.now().hour
        return when {
            hour < 12 -> "Good morning"
            hour < 17 -> "Good afternoon"
            else -> "Good evening"
        }
    }

    fun dashboardTagline(daysRemaining: Int, readiness: Int): String = when {
        daysRemaining == 0 -> "Appointment day — your file is ready to share with your doctor."
        daysRemaining <= 3 && readiness >= 70 -> "Almost there. Your doctor will have a clear picture."
        daysRemaining <= 3 -> "Every entry counts — $daysRemaining days until your appointment."
        readiness == 0 -> "Start today — your future self will thank you at the appointment."
        readiness < 40 -> "You're building something useful. Keep adding a little each day."
        readiness < 80 -> "Your file is taking shape. Steady progress."
        else -> "Strong file so far. Stay consistent until appointment day."
    }

    private fun isNoteEvent(event: TimelineEventResponse): Boolean =
        event.date_label.equals("Question", ignoreCase = true)

    private fun buildMissingHint(
        readiness: Int,
        daysRemaining: Int,
        actual: Int,
        expected: Int
    ): String? {
        val remaining = expected - actual
        return when {
            daysRemaining == 0 && readiness >= 80 -> null
            daysRemaining == 0 -> "A few more entries will strengthen your file for today."
            remaining <= 0 -> null
            readiness >= 90 -> "Almost complete — keep going."
            remaining == 1 -> "One more check-in fills a gap in your file."
            else -> "$remaining check-ins still to capture before your appointment."
        }
    }

    private fun computeStreak(measurements: List<TimelineEventResponse>): Int {
        if (measurements.isEmpty()) return 0
        val zone = ZoneId.systemDefault()
        val datesWithActivity = measurements.mapNotNull { event ->
            parseEventDate(event)?.atZone(zone)?.toLocalDate()
        }.toSet()

        var streak = 0
        var day = LocalDate.now()
        while (day in datesWithActivity) {
            streak++
            day = day.minusDays(1)
        }
        return streak
    }

    private fun parseEventDate(event: TimelineEventResponse): Instant? =
        runCatching {
            Instant.parse(event.effective_at ?: event.created_at)
        }.getOrNull()
}
